rewrite-core/src/main/java/org/openrewrite/SourceFile.java

```diff
@@ -36,7 +36,8 @@ public interface SourceFile extends Tree {
      * @return <code>true</code> if the parse-to-print loop is idempotent, <code>false</code> otherwise.
      */
     default boolean printEqualsInput(Parser.Input input, ExecutionContext ctx) {
-        return printAll().equals(StringUtils.readFully(input.getSource(ctx)));
+        Charset charset = (getCharset() == null) ? StandardCharsets.UTF_8 : getCharset();
+        return printAll().equals(StringUtils.readFully(input.getSource(ctx), charset));
```

Comment:
There's some slight subtleties that I want to point out here, and prevent those from causing issues by suggesting an alternative implementation.

The `StringUtils.readFully(InputStream)` has an early `if (inputStream == null) return "";`, that's not present in `readFully(InputStream, Charset)`. Not sure if we rely on that, but probably best to not change behavior just yet by calling the old method when there's no charset.

There's 14+ implementations of `SourceFile` in openrewrite/rewrite alone, with further implementations in modules for Kotlin, Python and JavaScript for instance. Some of these, such as Binary and Quark, throw an `UnsupportedOperationException` when `getCharset()` is called. I don't think the print equality is actually called for those, but figured call those out here all the same in case that triggers anyone to rethink the approach.

With the above I suggest we instead go for the following, to call `getCharset()` only once, after printing, and call the respective read operation that retains the `""` for a `null` `InputStream`.
```suggestion
        String printed = printAll();
        Charset charset = getCharset();
        if (charset != null) {
            return printed.equals(StringUtils.readFully(input.getSource(ctx), charset));
        }
        return printed.equals(StringUtils.readFully(input.getSource(ctx)));
```